File: hackerrank/easy/staircase/staircase.dart
import 'dart:io';

// Solution 1
// Time Complexity: O(n)
// Space Complexity: O(1)
void staircase(int n) {
  for (int index = 1; index <= n; index++) {
    stdout.write(" " * (n - index));
    stdout.write("#" * index);
    print("");
  }
}

// Solution 2
// Time Complexity: O(n2)
// Space Complexity: O(1)
void staircase2(int n) {
  for (int index = 1; index <= n; index++) {
    for (int i = 1; i <= (n - index); i++) {
      stdout.write(" ");
    }

    for (int i = 1; i <= index; i++) {
      stdout.write("#");
    }

    print("");
  }
}

void main() {
  int n = int.parse(stdin.readLineSync()!);
  staircase(n);
}

File: hackerrank/easy/staircase/staircase.swift

// Solution 1
// Time Complexity: O(n2)
// Space Complexity: O(n)
func staircase(n: Int) {
    for index in 1...n {
        let spaces = String(repeating: " ", count: n-index)
        let hashes = String(repeating: "#", count: index)
        print(spaces + hashes)
    }
}

// Solution 2
// Time Complexity: O(n2)
// Space Complexity: O(1)
func staircase2(n: Int) {
    for index in 1...n {
        for _ in 1...(n-index) {
            print(" ", terminator: "")
        }
        for _ in 1...index {
            print("#", terminator: "")
        }
        print()
    }
}